rust/connlib/tunnel/src/dns.rs

```diff
@@ -78,25 +78,40 @@ pub struct StubResolver {
     ip_provider: IpProvider,
     /// All DNS resources we know about, indexed by their domain (could be wildcard domain like `*.mycompany.com`).
     dns_resources: HashMap<String, ResourceDescriptionDns>,
+    /// Fixed dns name that will be resolved to fixed ip addrs, similar to /etc/hosts
+    hosts: HashSet<DomainName>,
+}
+
+fn fqdn_to_ips_for_hosts(hosts: &HashMap<String, Vec<IpAddr>>) -> HashMap<DomainName, Vec<IpAddr>> {
+    hosts
+        .iter()
+        .filter_map(|(d, a)| DomainName::vec_from_str(d).ok().map(|d| (d, a.clone())))
+        .collect()
+}
+
+fn ips_to_fqdn_for_hosts(hosts: &HashMap<String, Vec<IpAddr>>) -> HashMap<IpAddr, DomainName> {
+    hosts
+        .iter()
+        .filter_map(|(d, a)| {
+            DomainName::vec_from_str(d)
+                .ok()
+                .map(|d| a.iter().map(move |a| (*a, d.clone())))
+        })
+        .flatten()
+        .collect()
 }
 
-// We don't need to support multiple questions/qname in a single query because
-// nobody does it and since this run with each packet we want to squeeze as much optimization
-// as we can therefore we won't do it.
-//
-// See: https://stackoverflow.com/a/55093896
-/// Parses an incoming packet as a DNS query and decides how to respond to it
-///
-/// Returns:
-/// - `None` if the packet is not a valid DNS query destined for one of our sentinel resolvers
-/// - Otherwise, a strategy for responding to the query
 impl StubResolver {
-    pub(crate) fn new() -> StubResolver {
+    pub(crate) fn new(hosts: HashMap<String, Vec<IpAddr>>) -> StubResolver {
         StubResolver {
-            fqdn_to_ips: Default::default(),
-            ips_to_fqdn: Default::default(),
+            fqdn_to_ips: fqdn_to_ips_for_hosts(&hosts),
+            ips_to_fqdn: ips_to_fqdn_for_hosts(&hosts),
```

Comment:
Is this unclear? This works because then we hook this into the same system as the other records, only for these records we don't need the domain of the question to be a resource.

We could have these maps separated for hosts and have `get_records` take the maps by parameters and make it a free function. I want to parse the map instead of using it directly to not have to rewrite what is basically the `get_records` function for just this case and making the case for PTR records harder.

But I prefer to keep it this way since it requires less branching, so I think it's easier to understand.

Also, note here that if there are multiple records with the same ip addr it only affects `ips_to_fqdn` which is only used for PTR records, and it's okay to return only 1 of the records to correspond to the IP, that's how most DNS servers do it, normally there's only 1 PTR entry.

Reply:
There's an ugly edge case discvovered by the proptest here when a CIDR resource coincides with one of the IPS resolved by the known hosts.

Will separate these mappings.